### ISSUE TITLE
Weird script load in Firefox & Edge [dojo/b.js]

### ISSUE THREAD

**Issue by GeorgeKnap** [Esri/esri-loader#98]
### Expected behavior

- Describe what you expected or wanted to happen.
I want to async load arcgis javascript api 4.6 scripts into Angular 5 app. The arcgis scripts are hosted locally.

### Actual behavior
**In Chrome:** Works. As expected.
**In Firefox & Edge:** While loading scripts, esri-loader adds `<script>` tags into page header. That happens also in Chrome, but for some reason in Firefox & Edge it also tries to include `dojo/b.js` file. In Chrome there is no such file attempted to load, that's why in Chrome it works.
![esri-bullshit](https://user-images.githubusercontent.com/11920691/36844159-e37778b0-1d51-11e8-9d7a-96402bc48e68.PNG)

Obviously this file does not exists in arcgis javascript api and Firefox & Edge throws this errors:
![esri-bullshit-error](https://user-images.githubusercontent.com/11920691/36844321-8799de9c-1d52-11e8-9d92-a30acbf39be4.PNG)

### Steps to reproduce the behavior
this is how I'm loading scripts hosted locally. I tried both with loadScript() and loadModules() with configuration object passsed in:

```
async loadEsri() {
try {
                await loadScript({
                    url: '/assets/arcgis-js-api/4.6/dojo/dojo.js',
                    dojoConfig: {
                        async: true,
                        has: {
                            'esri-promise-compatibility': 1
                        }
                    }
                });
                let modules = await loadModules([
                     'esri/Map'
                     // ...more modules here...
                ]);
                this.Map = modules[0];
 }
catch (e) {
                console.error('ArcGIS: Error while downloading esri modules: ', e);
 }
}

```

this happens in cli  `ng serve` dev build. the build passes fine. error is at runtime.

**Comment by andygup** [comment]
You could very well be seeing an Angular component life-cycle issue. 

Here's a working Ionic `async/await` example: https://github.com/andygup/ionic2-esri-map/blob/master/pages/home/home.ts. I haven't tried it in stand-alone Angular or Angular CLI.

Ionic is a special case and in general we recommend simply waiting for `ngOnInit()` and loading `esri-loader` like this: https://github.com/Esri/angular-cli-esri-map/blob/master/src/app/esri-map/esri-map.component.ts#L65.

**Comment by GeorgeKnap** [comment]
Hi @andygup 
Thanks for sharing. I've redesigned the code to use simple `.then` callback like in your example but no luck.
Firefox and Edge are still trying to load `dojo/b.js` file.

I can't even find out which part of arcgis scripts  requires this script. I tried the full text search among all arcgis files but it's just not there.

**Comment by jwasilgeo** [comment]
How does it behave from the Esri CDN rather from a local JSAPI installation, using either asynchronous code syntax?

**Comment by andygup** [comment]
Adding to @jwasilgeo I also recommend testing your local arcgis api installation in a plain old JavaScript hello wold app.

**Comment by GeorgeKnap** [comment]
@jwasilgeo 
Alright. We are getting closer. This is what I have now and it works in all browsers!

```
async loadEsri() {
const config = {
                url: 'https://js.arcgis.com/4.6/',
                dojoConfig: {
                    async: true,
                    has: {
                        'esri-promise-compatibility': 1
                    }
                }
 };
loadModules([
    'esri/Map',
    //...more esri modules here...
  ], config).then((modules) => {
                  this.Map = modules[0];
            }).catch((e) => {
                console.error('ArcGIS: Error while downloading esri modules: ', e);
 });
```

So hosting files locally is the problem.
I downloaded the whole package from [Download page](https://developers.arcgis.com/javascript/latest/guide/get-api/index.html#download-api).

**Comment by davetimmins** [comment]
I had an issue with version 4.4 of the API hosted locally as the instructions said to change init.js but then reference dojo.js so that file needed changing too. Then I set the url to the dojo.js path e.g. arcgis_js_api/library/4.4/dojo/dojo.js

**Comment by GeorgeKnap** [comment]
Right. I've found another `baseUrl:"https://[HOSTNAME_AND_PATH_TO_JSAPI]dojo"` in `dojo.js` file which I've changed to point to existing dojo folder and I'm loading scripts via esri-loader with this config object:
```
{
                url: '/assets/arcgis-js-api/4.6/dojo/dojo.js',
                dojoConfig: {
                    async: true,
                    has: {
                        'esri-promise-compatibility': 1
                    }
                }
            }
```

Same result. Chrome works, Firefox & Edge tries to load `dojo/b.js` file.

**Comment by GeorgeKnap** [comment]
From what I've seen while debugging  `esri-loader` adds a `<script>` tag only to load a base script defined by config's `url` property.

In my case that is `<script type="text/javascript" src="/assets/arcgis-js-api/4.6/dojo/dojo.js" data-esri-loader="loaded"></script>`

after that debugging is not easy because dojo scripts are minified. But I can see that dojo replaces this tag with it's own `<script>` tags loading everything needed. In my case that is:
```
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/dojo/nls/dojo_cs.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/views/MapView.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/layers/MapImageLayer.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/ScaleBar.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/Expand.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/Print.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/geometry/geometryEngine.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/layers/GraphicsLayer.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/views/2d/draw/Draw.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/geometry/Circle.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/Fullscreen.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/moment/locale/cs.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/dojo/b.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/support/widget.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/Widget.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/ScaleBar/ScaleBarViewModel.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/Expand/ExpandViewModel.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/views/2d/draw/MultipointDrawAction.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/views/2d/draw/PointDrawAction.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/views/2d/draw/PolygonDrawAction.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/views/2d/draw/PolylineDrawAction.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/views/2d/draw/RectangleDrawAction.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/geometry/support/geodesicUtils.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/Fullscreen/FullscreenViewModel.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/ScaleBar/nls/ScaleBar.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/Fullscreen/nls/Fullscreen.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/views/nls/MapView_cs.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/support/widgetUtils.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/libs/maquette/maquette.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/Print/PrintViewModel.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/tasks/support/PrintTemplate.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/Print/FileLink.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/Print/TemplateOptions.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/views/2d/draw/input/DrawEvents.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/views/2d/draw/Keys.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/views/2d/draw/DrawAction.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/ScaleBar/nls/cs/ScaleBar.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/Fullscreen/nls/cs/Fullscreen.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/Print/nls/Print.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/tasks/PrintTask.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/tasks/support/PrintParameters.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/widgets/Print/nls/cs/Print.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/dojox/gfx/canvas.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/tasks/Geoprocessor.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/tasks/support/printTaskUtils.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/dojox/gfx/shape.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/dojox/gfx/path.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/dojox/gfx/arc.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/dojox/gfx/decompose.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/dojox/gfx/bezierutils.js"></script><script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/layers/support/MapImage.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/tasks/support/DataFile.js"></script><script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/tasks/support/Date.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/tasks/support/GPMessage.js"></script><script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/tasks/support/JobInfo.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/tasks/support/LinearUnit.js"></script><script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/tasks/support/ParameterValue.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/tasks/support/RasterData.js">
```

The question is why dojo adds this script?
`<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/dojo/b.js"></script>`

**Comment by tomwayson** [comment]
TLDR: that's how Dojo's AMD loader works.

If you create a simple HTML page and manually add a `<script>` tag w/ the `dojoConfig`, then add  `<script type="text/javascript" src="/assets/arcgis-js-api/4.6/dojo/dojo.js" data-esri-loader="loaded"></script>`, then put a `<script>` tag w/ a `require()` call that does whatever you're trying to do in `loadModules()`, do you see any different behavior?

> dojo replaces this tag with it's own <script> tags loading everything needed

Are you sure it _replaces_ it w/ those scripts, or does it just inject the additional `<script>` tags?

**Comment by GeorgeKnap** [comment]
@tomwayson 

> Are you sure it replaces it w/ those scripts, or does it just inject the additional <script> tags?

Replaces. I've just double checked.

I tried what you've suggested and this is the result:
```
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/dojo/nls/dojo_cs.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/dojo/cldr/nls/number.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/dojo/cldr/nls/gregorian.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/nls/common.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/nls/basemaps.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/dojo/cldr/nls/cs/number.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/dojo/cldr/nls/cs/gregorian.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/nls/cs/common.js"></script>
<script type="text/javascript" charset="utf-8" src="http://localhost:8080/assets/arcgis-js-api/4.6/esri/nls/cs/basemaps.js">

    <script>
        window.dojoConfig = {
            async: true,
            baseUrl: "http://localhost:8080/assets/arcgis-js-api/4.6/dojo",
            has: {
                "esri-promise-compatibility": 1
            }
        };
    </script>
    <script type="text/javascript" src="/assets/arcgis-js-api/4.6/dojo/dojo.js" data-esri-loader="loaded"></script>
    <script type="text/javascript">
        require(["esri/Map"], function (Map) {
            console.log("MAP LOADED:", Map);
        });
    </script>
```

the code worked. `console.log` was successful. No `dojo/b.js` loaded. Also got these two errors:
![vystrizek](https://user-images.githubusercontent.com/11920691/36962175-c8b16054-204e-11e8-9839-aea903a656cf.PNG)

**Comment by GeorgeKnap** [comment]
I have not thought this could be relevant but I was also loading external maps API library with 
`<script type="text/javascript" src="https://api.mapy.cz/loader.js"></script>`

Dojo has freaked out somehow and caused the behavior described above.
I fixed that by loading this library using dojo aliases like this:
```
 const config = {
                url: '/assets/arcgis-js-api/4.6/init.js',
                dojoConfig: {
                    async: true,
                    baseUrl: this.configService.get(ConfigProperty.EsriBaseUrl),
                    has: {
                        'esri-promise-compatibility': 1
                    },
                    aliases: [
                        ['mapy.cz', 'https://api.mapy.cz/loader.js']
                    ]
                }
            };
            loadModules([
                'esri/Map',
                'mapy.cz'
            ], config).then((modules) => {
...
}
```

Now Chrome & Firefix works. 
IE is still throwing the missing `dojo/b.js` and `multipleDefine` errors but the app works.

Hopefully it helps someone else running into these errors.
